src/core/mount-setup.c

```diff
@@ -397,9 +397,19 @@ int mount_setup(bool loaded_policy) {
          * nspawn and the container tools work out of the box. If
          * specific setups need other settings they can reset the
          * propagation mode to private if needed. */
-        if (detect_container() <= 0)
-                if (mount(NULL, "/", NULL, MS_REC|MS_SHARED, NULL) < 0)
-                        log_warning_errno(errno, "Failed to set up the root directory for shared mount propagation: %m");
+        if (mount(NULL, "/", NULL, MS_REC|MS_SHARED, NULL) < 0)
+                log_warning_errno(errno, "Failed to set up the root directory for shared mount propagation: %m");
+
```

Comment:
> Special setups that want private mounts can still run mount --make-rprivate (MS_PRIVATE|MS_REC) from within the container to undo the MS_SHARED, so I think the comment is still correct.

Hmm, no. IMO code + comment clearly says that from kernel to systemd we undo private, but we don't undo later any other setup, especially ones that broadcast we are containers, so this is a regression and as said we have to find a way to hack around it ... 

> I don't follow the security issue: when this function mount_setup() is executed by systemd as the pid1 of a container, it does not share the mounts inside with the mounts outside.
And unbindable mounts can be re-mounted without the unbindable bit with CAP_SYS_ADMIN in the container. And since systemd needs CAP_SYS_ADMIN to run anyway, this unbindable feature doesn't protect against systemd as the pid1 of a container.
What integration would you like to do? Are you preparing some unbindable mounts before calling systemd-nspawn?

I'm not referring to mounts outside, but mounts and directories inside that can be private for a specific purpose, and I'm trying to use systemd as pid1 in another setup, not nspawn.